### ISSUE TITLE
Surface reconstruction

### ISSUE THREAD

**Issue by ThibaultGROUEIX** [bearprin/neuralpull-pytorch#2]
Hi,

I have just tried the code! Thanks a lot for the pytorch version, it is very helpful.

This is the output I get when I fit NeuralPull on the horse for surface reconstruction. There are strange artefacts. 

![Screenshot 2021-09-29 at 23 50 22](https://user-images.githubusercontent.com/11445067/136092833-3109ca5e-2cfb-4fdf-aaf0-48acb33abf20.png)

Have you experienced the same thing on your end and do you have a working example?

Best regards,
Thibault

**Comment by bearprin** [comment]
> Hi,
> 
> I have just tried the code! Thanks a lot for the pytorch version, it is very helpful.
> 
> This is the output I get when I fit NeuralPull on the horse for surface reconstruction. There are strange artefacts.
> 
> ![Screenshot 2021-09-29 at 23 50 22](https://user-images.githubusercontent.com/11445067/136092833-3109ca5e-2cfb-4fdf-aaf0-48acb33abf20.png)
> 
> Have you experienced the same thing on your end and do you have a working example?
> 
> Best regards, Thibault

- I suggest fitting one shape by one network though the Implementation support fitting several shapes by condition code(feat). This will reduce the difficulty of convergence. Just put only one .npy in the 'npy' folder.
- You can use more epochs for fitting and then output the reconstruction results for each epoch to check if that network convergence.

Below is my horse (From [Points2Surf](https://github.com/ErlerPhilipp/points2surf) dataset) reconstruction result after the 207 epochs.

![horse](https://user-images.githubusercontent.com/90443362/136146546-f13ec5ed-7406-4ee2-beb1-ca1cb292287c.png)

 Hope help.

**Comment by ThibaultGROUEIX** [comment]
Thanks @bearprin this is great!
Let's try to get me the same results.

* I was already following suggestion 1, fitting a single shape by only having the horse in the `npy` folder
* I think suggestion 2 is natively in the code, at least it returns a mesh for every epoch off-the-shelf. 
My horse after the same number of epochs does not look good : the `L1 Loss is around 0.12`.

Did you keep the default parameters? 
```
parser.add_argument('--bd', type=float, default=0.55)
parser.add_argument('--resolution', type=int, default=128)
parser.add_argument('--thresholds', type=list, default=[0.0])

parser.add_argument('--name', type=str, default='base')
parser.add_argument('--epochs', type=int, default=1000)
parser.add_argument('--batch_size', type=int, default=16)
parser.add_argument('--lr', type=float, default=0.0001)
parser.add_argument('--num_workers', type=int, default=20)
parser.add_argument('--seed', type=int, default=40938661)
```

On my end, I followed the `quickstart section` instructions. The only modification i made was to remove the `fandisk` from the `npy` folder to only keep the horse before running `python train.py`
This is confusing me :)

**Comment by bearprin** [comment]
I have made some modifications to the code.

- deleted the fandisk shape(mesh and point cloud).
- Besides, I have repair some issues of env.yaml.

After that, I retrained the network for 'horse' with the default hyperparameters. Below is my result after 212 epochs.

![horse_212](https://user-images.githubusercontent.com/90443362/136220765-a2a238e7-46dd-4bb0-859e-255811a16c9b.png)

**Try reclone this repository with the latest code and retrain again.**


> Thanks @bearprin this is great! Let's try to get me the same results.
> 
> * I was already following suggestion 1, fitting a single shape by only having the horse in the `npy` folder
> * I think suggestion 2 is natively in the code, at least it returns a mesh for every epoch off-the-shelf.
>   My horse after the same number of epochs does not look good : the `L1 Loss is around 0.12`.
> 
> Did you keep the default parameters?
> 
> ```
> parser.add_argument('--bd', type=float, default=0.55)
> parser.add_argument('--resolution', type=int, default=128)
> parser.add_argument('--thresholds', type=list, default=[0.0])
> 
> parser.add_argument('--name', type=str, default='base')
> parser.add_argument('--epochs', type=int, default=1000)
> parser.add_argument('--batch_size', type=int, default=16)
> parser.add_argument('--lr', type=float, default=0.0001)
> parser.add_argument('--num_workers', type=int, default=20)
> parser.add_argument('--seed', type=int, default=40938661)
> ```
> 
> On my end, I followed the `quickstart section` instructions. The only modification i made was to remove the `fandisk` from the `npy` folder to only keep the horse before running `python train.py` This is confusing me :)

**Comment by ThibaultGROUEIX** [comment]
Huge thanks ! Trying now.

**Comment by ThibaultGROUEIX** [comment]
Interesting. So I restarted from scratch, reinstalled the conda env and launched. Here is my Tensorboard log.

As you can see, the test losses are very unstable. 
![Screenshot 2021-10-06 at 17 33 32](https://user-images.githubusercontent.com/11445067/136236091-8905d3d4-e047-465f-b04a-fe5cf0791971.png)


After 500 epochs, I still don't have a great horse like you. If i pick the epoch with the best L1 loss, I get this:
![Screenshot 2021-10-06 at 17 38 33](https://user-images.githubusercontent.com/11445067/136236708-b7a9c145-4d5c-46f0-8780-83cfd439a3ad.png)

I am not sure how to address this. Do you also observe pretty bad convergence properties in your curves (ideally can you share them?)?

**Comment by mabaorui** [comment]
Thank you for the pytorch version. But this line of code is slightly different from the original work, which may lead to incomplete structure. 
@ThibaultGROUEIX You can try to change it to [0.005] or [0.0,0.0005,0.005], and then observe whether the output mesh is normal.
Finally, thank you very much for your work. @bearprin 
![image](https://user-images.githubusercontent.com/14806520/136330206-5ee5eab8-842a-40b5-8f72-ec20082166dc.png)

**Comment by bearprin** [comment]
> Thank you for the pytorch version. But this line of code is slightly different from the original work, which may lead to incomplete structure. @ThibaultGROUEIX You can try to change it to [0.005] or [0.0,0.0005,0.005], and then observe whether the output mesh is normal. Finally, thank you very much for your work. @bearprin ![image](https://user-images.githubusercontent.com/14806520/136330206-5ee5eab8-842a-40b5-8f72-ec20082166dc.png)

Sorry for this mistake, and I repaired that.

**Comment by ThibaultGROUEIX** [comment]
Hi ! Thanks a lot @mabaorui and @bearprin,

The `0.005` helped get this results :
![Screenshot 2021-10-07 at 11 50 36](https://user-images.githubusercontent.com/11445067/136361963-ab353141-5ad3-4327-9415-f4b5eaa6f5d8.png)

However, immediately after I get this: 
![Screenshot 2021-10-07 at 11 49 25](https://user-images.githubusercontent.com/11445067/136362286-39dc0111-e36d-4008-b051-d5502327382e.png)

Do you also observe some instability in the results, and is it expected that one has to look for good L1 test values to find a good reconstruction, or all reconstructions after convergence should be good?

To help inform this, can you share your Tensorboard curves @bearprin ?

------------

The programmer in me is still struggling with the fact that we have the same conda env, an experiment with a fixed seed, and @bearprin gets good results and I don't (with the 0 threshold). 
One thing I noticed is that knn_cuda is not in the yaml spec file, I figured it came from https://github.com/unlimblue/KNN_CUDA Do you confirm @bearprin ?
On which platform are you running the code @bearprin ?

Thanks a lot for the time you take on this,
Thibault

**Comment by bearprin** [comment]
> Hi ! Thanks a lot @mabaorui and @bearprin,
> 
> The `0.005` helped get this results : ![Screenshot 2021-10-07 at 11 50 36](https://user-images.githubusercontent.com/11445067/136361963-ab353141-5ad3-4327-9415-f4b5eaa6f5d8.png)
> 
> However, immediately after I get this: ![Screenshot 2021-10-07 at 11 49 25](https://user-images.githubusercontent.com/11445067/136362286-39dc0111-e36d-4008-b051-d5502327382e.png)
> 
> Do you also observe some instability in the results, and is it expected that one has to look for good L1 test values to find a good reconstruction, or all reconstructions after convergence should be good?
> 
> To help inform this, can you share your Tensorboard curves @bearprin ?
> 
> The programmer in me is still struggling with the fact that we have the same conda env, an experiment with a fixed seed, and @bearprin gets good results and I don't (with the 0 threshold). One thing I noticed is that knn_cuda is not in the yaml spec file, I figured it came from https://github.com/unlimblue/KNN_CUDA Do you confirm @bearprin ? On which platform are you running the code @bearprin ?
> 
> Thanks a lot for the time you take on this, Thibault

The latest eny.yaml has KNN_CUDA, see [here](https://github.com/bearprin/neuralpull-pytorch/blob/master/env.yaml#L221).
